src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubAppCredentials.java

```diff
@@ -171,57 +221,239 @@ public String getUsername() {
-             clone.apiUri = apiUri;
-             clone.owner = owner;
-             clone.cachedToken = cachedToken;
-             clone.tokenCacheTime = tokenCacheTime;
-             return clone;
-         }
-
-     }
+        return new DelegatingGitHubAppCredentials(this);
+    }
+
+    private static final class DelegatingGitHubAppCredentials extends BaseStandardCredentials implements StandardUsernamePasswordCredentials {
+
+        private static final String SEP = "%%%";
+
+        private final String appID;
+        /** 
+         * An encrypted form of all data needed to refresh the token.
+         * Used to prevent {@link GetToken} from being abused by compromised build agents.
+         */
+        private final String tokenRefreshData;
+        private AppInstallationToken cachedToken;
+
+        private transient Channel ch;
+
+        DelegatingGitHubAppCredentials(GitHubAppCredentials onMaster) {
+            super(onMaster.getScope(), onMaster.getId(), onMaster.getDescription());
+            JenkinsJVM.checkJenkinsJVM();
+            appID = onMaster.appID;
+            tokenRefreshData = Secret.fromString(onMaster.appID + SEP + onMaster.privateKey.getPlainText() + SEP + onMaster.actualApiUri() + SEP + onMaster.owner).getEncryptedValue();
+
+            // Check token is valid before sending it to the agent.
+            // Ensuring the cached token is not stale before sending it to agents keeps agents from having to
+            // immediately refresh the token.
+            // This is intentionally only a best-effort attempt.
+            // If this fails, the agent will fallback to making the request (which may or may not fail).
+            try {
+                LOGGER.log(Level.FINEST, "Checking App Installation Token for app ID {0} before sending to agent", onMaster.appID);
+                onMaster.getPassword();
+            } catch (Exception e) {
+                LOGGER.log(Level.WARNING, "Failed to refresh stale GitHub App installation token before sending to agent for app ID " + onMaster.getAppID(), e);
+            }
+
+            cachedToken = onMaster.getCachedToken();
+        }
+
+        private Object readResolve() {
+            JenkinsJVM.checkNotJenkinsJVM();
+            synchronized (this) {
+                ch = Channel.currentOrFail();
+            }
+            return this;
+        }
+
+        @NonNull
+        @Override
+        public String getUsername() {
+            return appID;
+        }
+
+        @Override
+        public Secret getPassword() {
+            JenkinsJVM.checkNotJenkinsJVM();
+            try {
+                String appInstallationToken;
+                synchronized (this) {
+                    try {
+                        if (cachedToken == null || cachedToken.isStale()) {
+                            cachedToken = ch.call(new GetToken(tokenRefreshData));
+                            LOGGER.log(Level.FINER,
```

Comment:
@dwnusbaum  @jglick 
I'd like to be able to check this output in the test, but it doesn't show up in my test logging handler.  Only the logs from the controller show up.  I know there's a way, I just don't know what it is.

Reply:
Okay, I think I see something that looks like the right thing.
https://github.com/jenkinsci/jenkins/blob/742c99eab9c878ceca6ae5171eaaf3b892ee8d0e/core/src/main/java/hudson/logging/LogRecorder.java

Reply:
Figured it out.